feat: add dynamic collection type ID (#17)

mx42 opened the pull request:
Hello,

Related to the feature request I opened [a couple of days ago](https://github.com/rclsilver/home-assistant-ecocito/issues/17), here is some proposed implementation.

* Added a method to the Ecocito client to fetch collection type IDs.
* Added an option-flow to be able to configure the integration type ID mapping after setup, pulling the types via the previous method.
* Changed the coordinators and clients implementation to fetch the configured types instead (fallback to the hard-coded default so shouldn't change the behavior on non-configured instances).
* Since the garbage and recycle implementation became identical, merging of some methods.
* Added a configuration for the refresh time, since the value should change at most once per day (more realistically maybe once per week).

Future idea: probably instead of hard-coding garbage collection + recycling collection, maybe the user could just select some custom sensors (we could probably just map the types we fetch from the site to a sensor name or none for deactivation).

Please tell me what you think.

mx42 committed:
chore: add nix flake

mx42 committed:
feat: add a method to dynamically fetch available collection type ids

mx42 committed:
chore: add some constants for exception messages to satisfy the linter

mx42 committed:
feat: add an option-flow to configure type ID mappings (unused)

mx42 committed:
feat: refactor the coordinator and client to dynamically get and use the type ID from config

mx42 committed:
chore: update ruff to 0.11.0

mx42 committed:
feat: merge coordinators to have less duplicate code

mx42 committed:
chore: update ruff rules

rclsilver commented:
Hi @mx42! Thanks for your pull request, and sorry for not getting back to you sooner... I'll try to test your work quickly so I can merge your request.

rclsilver commented:
Hi @mx42,

I just installed your version in my dev container and when I try to configure the extension, I have the following stack trace:
```
Config flow could not be loaded: 500 Internal Server Error Traceback (most recent call last): File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py", line 452, in _handle_request resp = await request_handler(request) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_app.py", line 543, in _handle resp = await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_middlewares.py", line 114, in impl return await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware return await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/forwarded.py", line 83, in forwarded_middleware return await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/request_context.py", line 26, in request_context_middleware return await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/ban.py", line 85, in ban_middleware return await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/auth.py", line 242, in auth_middleware return await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/headers.py", line 32, in headers_middleware response = await handler(request) ^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/helpers/http.py", line 73, in handle result = await handler(request, **request.match_info) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/decorators.py", line 81, in with_admin return await func(self, request, *args, **kwargs) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/config/config_entries.py", line 262, in post return await super().post(request) ^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/components/http/data_validator.py", line 70, in wrapper return await method(view, request, data, *args, **kwargs) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/helpers/data_entry_flow.py", line 77, in post return await self._post_impl(request, data) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/helpers/data_entry_flow.py", line 84, in _post_impl result = await self._flow_mgr.async_init( ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/data_entry_flow.py", line 333, in async_init flow = await self.async_create_flow(handler, context=context, data=data) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/config_entries.py", line 2543, in async_create_flow return handler.async_get_options_flow(entry) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/workspaces/ecocito/custom_components/ecocito/config_flow.py", line 73, in async_get_options_flow from .options_flow import EcocitoOptionsFlowHandler File "/workspaces/ecocito/custom_components/ecocito/options_flow.py", line 8, in <module> from homeassistant import ConfigFlowResult, config_entries ImportError: cannot import name 'ConfigFlowResult' from 'homeassistant' (/home/vscode/.local/ha-venv/lib/python3.12/site-packages/homeassistant/__init__.py)
```

mx42 commented:
Ah, I didn't change the `homeassistant` version in requirements.txt, I directly tested on my home instance (running the latest HA).
I'll check it out later tonight. Sorry for this...

mx42 committed:
chore: bump home-assistant req to latest

mx42 commented:
Hello, can you try with the updated requirements.txt (with latest `homeassistant` package version) ?
In the meantime I'll try to install vscode to test the devcontainer.

EDIT: Ok one change I missed was the need to upgrade to Python 3.13. I've pushed the update on `Dockerfile.dev`. Somehow the dev container seem stuck on my end, but it should be better, I guess?

mx42 committed:
chore: update Python version to 3.13 for the latest HA

rclsilver commented:
Hi, I still have some troubles with your branch. Any chance to contact you by an IM? I'm available on Telegram and Signal

mx42 commented:
> Hi, I still have some troubles with your branch. Any chance to contact you by an IM? I'm available on Telegram and Signal

Sure, I shot you an email.

mx42 committed:
fix: fix broken imports

mx42 committed:
feat: make the sync of garbage types optional

If the selected type is "Skip synchro", the ecocito data won't contain the related coordinator(s)
and any sync of that type will be skipped.

mx42 committed:
feat: grab all collection events in 1 call and dispatch in several sensors after

It becomes a bit of a mess with a lot of dynamically named sensors, and possibly a lot of empty ones.
It adds a bit of a burden on the user to clean-up the mess. :/
